### PR TITLE
Lab 1 - Fib

### PR DESCRIPTION
**Pull request by NicholeL**
Fixes #579 

Review of colleague's PR 

Changes proposed in this PR:
- Implement Lab1 - Fibonacci series

### CODE REVIEW

File: 01_fib/NicholeL/fib.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func abs(args int) (res int) {
```
Comment (juliaogris) [inline 01_fib/NicholeL/fib.go]: `arg` (singular) or `val` no need for name return, it is discouraged unless you have good reason (see [Effective Go](https://golang.org/doc/effective_go.html#named-results) and [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters)).

File: 01_fib/NicholeL/fib.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	if args == 0 {
+		return 0
+	}
+	if args > 0 {
```
Comment (juliaogris) [inline 01_fib/NicholeL/fib.go]: get rid of first three lines and change to `>=`

File: 01_fib/NicholeL/fib.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+func fib(args int) {
+	a, b := 0, 1
+	if abs(args) > 92 {
+		fmt.Fprint(out, "Fibonacci numbers greater than 92 not supported")
+		return
+	}
+	if abs(args) <= 1 {
+		fmt.Fprintln(out, args)
```
Comment (juliaogris) [inline 01_fib/NicholeL/fib.go]: return here rather than `if`.
Deep nesting levels are harder to comprehend.

File: 01_fib/NicholeL/fib.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+		return 0
+	}
+	if args > 0 {
+		return args
+	}
+	return args * (-1)
+
+}
+
+func fib(args int) {
```
Comment (juliaogris) [inline 01_fib/NicholeL/fib.go]: use plural only for slices/arrays and maybe maps.
change to arg or val

File: 01_fib/NicholeL/fib.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	a, b := 0, 1
+	if abs(args) > 92 {
```
Comment (juliaogris) [inline 01_fib/NicholeL/fib.go]: this is only true if you host system defaults to int64 for int (check the go [playground](https://play.golang.org) for a 32 bit int default).
Use explicit int64 or get rid of check (not required by lab assignment).

File: 01_fib/NicholeL/fib_test.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+13
+`},
+		{99, `Fibonacci numbers greater than 92 not supported`},
+	}
+	for _, testCase := range tests {
+		var buf bytes.Buffer
+		out = &buf
+		fib(testCase.input)
+		excepted := strconv.Quote(testCase.excepted)
+		actual := strconv.Quote(buf.String())
```
Comment (juliaogris) [inline 01_fib/NicholeL/fib_test.go]: remove `Quote` on both.

File: 01_fib/NicholeL/fib.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+				res := a + b
+				a, b = b, res
+				if args < 0 && index%2 == 0 {
+					fmt.Fprintln(out, res*(-1))
+				} else {
+					fmt.Fprintln(out, res)
+				}
+			}
+		}
+	}
```
Comment (juliaogris) [inline 01_fib/NicholeL/fib.go:36]: This function seems a bit more complicated than necessary. Here parts of our reference solution:

```go
func fib(n int) {
	sign := // 1 for positive n ; -1 for negative n; define inline
	n1, n2 := 0, 1
	for j := 0; j < n*sign; j++ {
		n1, n2 = n2, n1+sign*n2
		// print n1
	}
}
```

Feel free to take, ignore or re-work.

**Review by juliaogris — CHANGES_REQUESTED** [review]
Please read through the [Collected Code Review Comments](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments)
- [Fix your pr description](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#pr-description)
- [Fix commit messages](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#commit-messages)

I've also left a few comments with the code.

File: 01_fib/NicholeL/fib.go (hunk 1 above)
Comment (NicholeL) [reply 01_fib/NicholeL/fib.go]: ok, thanks for juliaogris's reminder

File: 01_fib/NicholeL/fib.go (hunk 2 above)
Comment (NicholeL) [reply 01_fib/NicholeL/fib.go]: will update

File: 01_fib/NicholeL/fib.go (hunk 3 above)
Comment (NicholeL) [reply 01_fib/NicholeL/fib.go]: yep, thanks for your advice, will update to return

File: 01_fib/NicholeL/fib.go (hunk 4 above)
Comment (NicholeL) [reply 01_fib/NicholeL/fib.go]: will update

File: 01_fib/NicholeL/fib.go (hunk 5 above)
Comment (NicholeL) [reply 01_fib/NicholeL/fib.go]: will delete this check

File: 01_fib/NicholeL/fib_test.go (hunk 6 above)
Comment (NicholeL) [reply 01_fib/NicholeL/fib_test.go]: will remove Quote

File: 01_fib/NicholeL/fib.go (hunk 7 above)
Comment (NicholeL) [reply 01_fib/NicholeL/fib.go:36]: thanks for your suggestion

**Review by anzboi — CHANGES_REQUESTED** [review]
LGTM
You just need to link a colleagues PR that you have reviewed

**Pull request closed without merging**